### ISSUE TITLE
Reorganize how/when we save cropped images

### ISSUE THREAD

**Issue by misaugstad** [ProjectSidewalk/SidewalkWebpage#3468]
##### Brief description of problem/feature
@hoominchu recently added a feature that saves cropped images of the GSV window when a user adds a label. Right now, these are added through a dedicated /saveImage endpoint. Instead of being it's own endpoint, I'd like to log this with all the other data we log. The benefits of this are:
1. I think that it just fits the model that we're using elsewhere. I would think of these cropped images as essentially metadata about a label, so we might as well submit it with the rest of the metadata for the labels.
2. When we don't submit the images at the same time of the labels, we don't know the labels' label IDs! Once the metadata is submitted to the server, a label ID is generated. So instead of saving the cropped image with the label ID, we have to use the "temporary_label_id" and "user_id" to uniquely identify labels. And in the future we'd need to write a script to go and update the filenames to use the "label_id" instead... We could avoid this whole process if we save the crop at the same time, because we would have access to the "label_id"!

Some considerations:
1. We should only do this if it doesn't negatively impact performance. The nice part about what we're doing right now is that the image can be converted to b64 async, and the larger payload doesn't slow down our logging. Thankfully, we are rarely waiting on a response from our logging for any reason on the Explore page, it's almost entirely async. The only thing we wait for is when the user finishes a mission. BUT, a mission is finished when a user steps forward. And I would think that the time between placing a label and taking a step forward would hopefully be enough to convert the image to b64. We should do ample testing to make sure, though!
2. After we make this change, we should then write a one-time script to go through and update the filenames of the cropped images that we've already collected, replacing the "temporary_label_id" and "user_id" with the single "label_id" from the db.
3. The current flow: every time we log something, we check if there are at least 200 logs stored up. If so, we send those logs to the server. I think that we should just change it so that it's if there are at least 200 logs stored _and_ any image crops we want to send are ready to go.
4. The only complicated part is what do we want to do if we try to send data because a mission ended, but we haven't finished converting an image to b64? Do we fire an event any time an image is ready to be sent? Maybe the whole logging system should be event-based like this, and all the logic for whether or not to send data is in the event callback.
5. If the cropping fails for whatever reason, we should log that. And we should make sure that if cropping fails, it doesn't prevent us from ever sending data.

**Comment by jonfroehlich** [comment]
Nice thinking @misaugstad! Are you saying that we'll save these as blobs in the database or still in the filesystem?

**Comment by misaugstad** [comment]
Still in the filesystem! It's just about when we transfer from the front-end to the server. I want to do it at the same time that we are logging everything else instead of doing it separately!

**Comment by misaugstad** [comment]
An alternative if we wanted to keep this as a separate endpoint: We could wait to send the crops until we get a response from the server with the label's actual `label_id`. Then we know how to name the file and don't have to make those changes in the future. I'd like to avoid waiting to do it, but if it causes performance issues to do it at the same time, I think that waiting for that response from the server before sending is reasonable.

**Comment by jonfroehlich** [comment]
I’d totally defer to you on this one Mikey. Happy to discuss it further if
you want a sounding board (Minchu could be good for this too)

**Comment by misaugstad** [comment]
Sounds good! A last thought I want to record: as we talk about in #3469, the actual I/O to save the image should happen async on the server-side if we include it with the rest of the logging. We already have something else set up to happen async when data is logged (sending data to scistarter), so this should be done in the same vein so that we don't hold up sending a response back to the client.

**Comment by misaugstad** [comment]
TL;DR: I think that we should
1. send the image crops async _after_ the label is sent and we get the `label_id` in response, and
2. start sending logs every time a label is placed to reduce the likelihood that we fail to save an image

>An alternative if we wanted to keep this as a separate endpoint: We could wait to send the crops until we get a response from the server with the label's actual `label_id`. Then we know how to name the file and don't have to make those changes in the future. I'd like to avoid waiting to do it, but if it causes performance issues to do it at the same time, I think that waiting for that response from the server before sending is reasonable.

Here I said that I'd prefer to send the image at the same time as the rest of the data unless it causes performance issues. I now do believe that this would impact the user, specifically when it comes to how long they wait before being able to click the "Continue" button to start a new mission.

Each image is around 5-10 megabits in size, and I believe that 10-30 megabits/second are very common upload speeds (would expect worse outside the US, on public wifi, etc). So this would sometimes mean an additional whole second of waiting time before users can click on the Continue button.

It also means that for users on a slow network, data logging is going to be received out of order because a batch that is sent later can be received earlier than a batch that is sent earlier that contained an image. We should theoretically be dealing with this case correctly already, but there is a ton of stuff that we're recording with this endpoint, and it would be easy to make a mistake. I could think of some ideas for safeguards to put in place for this, but it could mean some bigger changes/refactoring. Depends on how important we think it is.

So I'm suggesting that we instead continue to send the images async (like we are doing now), but we wait to send them until we've submitted the label to the server and have gotten back the label_id in response. The downside: there are edge cases where I could see us failing to save an image crop. Specifically, if someone adds a label and then quickly closes the tab or refreshes the page. We have something set up to quickly send off a last batch of logged data if the page closes/refreshes, but we don't have the ability to send data, wait for a response, and then send more data before the page closes.

I believe that that should be quite rare; if a user moves their mouse around to add a severity and tags before closing the page, this shouldn't be an issue. I think that we should also just send a batch of logs any time a label is placed, further reducing the likelihood that someone could close the page before we send the image. This could also lead to more instances of logs showing up out of order, but probably not a big deal.

**Comment by jonfroehlich** [comment]
I think this seems reasonable. Thanks for your deep thinking about this.

And to be clear, this doesn't chang when we capture the image, just when we upload it?

**Comment by misaugstad** [comment]
>And to be clear, this doesn't chang when we capture the image, just when we upload it?

Correct!

**Comment by misaugstad** [comment]
Reopening this until after we run the script to rename all the files that used the old naming convention on the production servers.

**Comment by misaugstad** [comment]
All the crops on the server have been fixed, all done here!
